lewis/adapters/stream.py

```diff
@@ -163,8 +186,12 @@ def __init__(self, func, pattern, argument_mappings=None, return_mapping=None, d
             raise RuntimeError('Can not construct a Func-object from a non callable object.')
 
         self.func = func
-        self.raw_pattern = pattern
-        self.pattern = re.compile(b(pattern), 0) if pattern else None
+
+        self.raw_pattern = pattern.raw_pattern
+        self.pattern = re.compile(b(pattern.regex_pattern), 0)
```

Comment:
Some thoughts / explorations on first point:

1. All our code currently begins all regexes with `^`.
1. I would say that if we implicitly match (or the device developer implicitly describes) expected TCP Stream commands _not_ starting with the beginning of the string, this would be a logic error. Because if you search for a match of `L?` _anywhere_ in the incoming TCP Stream command, that will equally match incoming `L?`, `notL?` and `MRGLGLGLGL?` commands. 
The last two are probably not intended to match (and should therefore cause an error response). 
If they are intended to match, I think it should be _required_ to _explicitly_ write the `pattern` to match that (or anything `.*L?`), so that it is immediately visible that this is intended when looking at the `pattern`. It should not be assumed that this is intended or desired just because the `^` is missing.
1. Testing some things in a Python console just now I ran into a surprise: 
**`re.match` assumes the `^` behaviour by default!!** (but not `$`)
So even if you do this: `Cmd('get_status', regex(r'S\?'))`, this will match `S?` and `S?foobar` BUT NOT `foobarS?`.
1. Snooping around the IBEX code a bit, I saw that almost all of their regexes start with `^`. The only excepting is this: https://github.com/ISISComputingGroup/EPICS-DeviceEmulator/blob/afccb520b9c8e60b21b84e5ef6cc42c66c3e8ec1/lewis_emulators/neocera_ltc21/interfaces/stream_interface.py
Which is very interesting indeed.

I agree about there perhaps being a third option to consider... need to think about this some more.